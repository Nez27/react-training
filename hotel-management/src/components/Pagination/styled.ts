import styled from 'styled-components';

interface IPaginationBtn {
  disabled: boolean;
}

const StyledPagination = styled.div`
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: space-between;
  font-size: var(--fs-sm-x);
`;

const Buttons = styled.div`
  display: flex;
  gap: 30px;
`;

const PaginationBtn = styled.button<IPaginationBtn>`
  background-color: var(--secondary-btn-color);
  border: none;
  border-radius: var(--radius-sm);
  font-weight: 500;

  display: flex;
  align-items: center;
  justify-content: center;
  gap: 10px;

  padding: 5px 10px;

  transition: all 0.3s;

  & svg {
    height: 20px;
    width: 20px;
  }

  &:hover:not(:disabled) {
    background-color: var(--primary-color);
    color: var(--light-text);
  }

  &:disabled {
    cursor: not-allowed;
  }
`;

export { StyledPagination, PaginationBtn, Buttons };
