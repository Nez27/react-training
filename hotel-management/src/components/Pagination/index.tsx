import { DEFAULT_PAGE_SIZE } from '@constant/config';
import { useCallback, useMemo } from 'react';
import { useSearchParams } from 'react-router-dom';
import { Buttons, PaginationBtn, StyledPagination } from './styled';
import { MdKeyboardArrowRight, MdKeyboardArrowLeft } from 'react-icons/md';

interface IPagination {
  count: number;
}

const Pagination = ({ count }: IPagination) => {
  const [searchParams, setSearchParams] = useSearchParams();
  const currentPage = !searchParams.get('page')
    ? 1
    : Number(searchParams.get('page'));

  const totalPage = Math.ceil(count / DEFAULT_PAGE_SIZE);

  const nextPage = useCallback(() => {
    const next = currentPage === totalPage ? currentPage : currentPage + 1;

    searchParams.set('page', next.toString());
    setSearchParams(searchParams);
  }, [currentPage, searchParams, totalPage, setSearchParams]);

  const previousPage = useCallback(() => {
    const previous = currentPage === 1 ? currentPage : currentPage - 1;

    searchParams.set('page', previous.toString());
    setSearchParams(searchParams);
  }, [currentPage, searchParams, setSearchParams]);

  const fromIndex = useMemo(
    () => (currentPage - 1) * DEFAULT_PAGE_SIZE + 1,
    [currentPage]
  );

  const toIndex = useMemo(
    () => (currentPage === totalPage ? count : currentPage * DEFAULT_PAGE_SIZE),
    [currentPage, count, totalPage]
  );

  return (
    totalPage > 1 && (
      <StyledPagination>
        <p>
          Showing {fromIndex} to {toIndex} of <span>{count}</span> results
        </p>

        <Buttons>
          <PaginationBtn onClick={previousPage} disabled={currentPage === 1}>
            <MdKeyboardArrowLeft /> <span>Previous</span>
          </PaginationBtn>

          <PaginationBtn
            onClick={nextPage}
            disabled={currentPage === totalPage}
          >
            <span>Next</span>
            <MdKeyboardArrowRight />
          </PaginationBtn>
        </Buttons>
      </StyledPagination>
    )
  );
};

export default Pagination;
