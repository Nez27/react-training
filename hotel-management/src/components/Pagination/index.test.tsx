import { render, fireEvent, RenderResult } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import Pagination from '.';

describe('Pagination', () => {
  const count = 20;
  let wrapper: RenderResult | null = null;

  beforeEach(() => {
    wrapper = render(
      <BrowserRouter>
        <Pagination count={count} />
      </BrowserRouter>
    );
  });

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('Next button should be disable when in the last page', () => {
    fireEvent.click(wrapper!.getByText('Next'));

    expect(wrapper!.getByText('Next').closest('button')).toBeDisabled();
  });

  test('Previous button should not be disable when move to second page', () => {
    fireEvent.click(wrapper!.getByText('Next'));

    expect(wrapper!.getByText('Previous').closest('button')).not.toBeDisabled();
  });

  test('Previous button should be disabled when go back to the first page', () => {
    // Navigate to the second page
    fireEvent.click(wrapper!.getByText('Next'));

    // Back to the first page
    fireEvent.click(wrapper!.getByText('Previous'));

    expect(wrapper!.getByText('Previous').closest('button')).toBeDisabled();
  });
});
