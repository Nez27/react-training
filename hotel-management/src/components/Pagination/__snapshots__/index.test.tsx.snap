// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bdnyFh kjuCvI"
      >
        <p>
          Showing 
          1
           to 
          10
           of 
          <span>
            20
          </span>
           results
        </p>
        <div
          class="sc-gtsqUy gUbTlC"
        >
          <button
            class="sc-dlniIP kLAzMx"
            disabled=""
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
              />
            </svg>
             
            <span>
              Previous
            </span>
          </button>
          <button
            class="sc-dlniIP kLAzMx"
          >
            <span>
              Next
            </span>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0V0z"
                fill="none"
              />
              <path
                d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bdnyFh kjuCvI"
    >
      <p>
        Showing 
        1
         to 
        10
         of 
        <span>
          20
        </span>
         results
      </p>
      <div
        class="sc-gtsqUy gUbTlC"
      >
        <button
          class="sc-dlniIP kLAzMx"
          disabled=""
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
            />
          </svg>
           
          <span>
            Previous
          </span>
        </button>
        <button
          class="sc-dlniIP kLAzMx"
        >
          <span>
            Next
          </span>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
            <path
              d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
